llms/mlx_lm/utils.py

```diff
@@ -334,56 +341,85 @@ def generate(
        kwargs: The remaining options get passed to :func:`generate_step`.
           See :func:`generate_step` for more details.
     """
+    is_batch = isinstance(prompt, list)
     if not isinstance(tokenizer, TokenizerWrapper):
         tokenizer = TokenizerWrapper(tokenizer)
 
-    if verbose:
-        print("=" * 10)
-        print("Prompt:", prompt)
-
-    prompt_tokens = mx.array(tokenizer.encode(prompt))
-    detokenizer = tokenizer.detokenizer
+    if is_batch:
+        tokenizer._tokenizer.padding_side = "left"
```

Comment:
I see that we left pad shorter prompts here which makes sense. But one thing that I'm wondering is how this is handled in the causal models if at all? Shouldn't the causal mask take into account the padding?

Reply:
I didn't handle it, generation seems OK without it but indeed to be correct I should consume the `tokenizer._tokenizer(prompt, padding=True)["attention_mask"]`. To do this I would need to update our model APIs to have `attention_mask` as an input similar to how transformers has `model.generate` taking `attention_mask`. Probably this is involves hitting every file in `models/`. But shouldn't mostly be copy/paste. I can look into it.